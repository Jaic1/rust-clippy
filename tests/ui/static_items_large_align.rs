// The example is adapted from https://github.com/rust-lang/rust/issues/70022

#![warn(clippy::static_items_large_align)]

#[repr(align(0x100000))]
#[derive(Clone, Copy)]
struct Aligned(u8);

struct AlignedWrapper {
    f: u8,
    g: Aligned,
}

enum AlignedEnum {
    A(Aligned),
}

struct AlignedPtr(*const Aligned);

unsafe impl Sync for AlignedPtr {}

struct AlignedGeneric<T>(T);

static X: Aligned = Aligned(0);

static X_REF: &Aligned = &Aligned(0);

static X_PTR: AlignedPtr = AlignedPtr(&Aligned(0) as *const _);

static XW: AlignedWrapper = AlignedWrapper { f: 0, g: Aligned(0) };

static ARRAY: [Aligned; 10] = [Aligned(0); 10];

static TUPLE: (u8, (Aligned, u8)) = (0, (Aligned(0), 0));

static XE: AlignedEnum = AlignedEnum::A(Aligned(0));

static XG: AlignedGeneric<Aligned> = AlignedGeneric(Aligned(0));

static XGW: AlignedGeneric<AlignedWrapper> = AlignedGeneric(AlignedWrapper { f: 0, g: Aligned(0) });

fn main() {
    let x = Aligned(0);
    println!("{:#x}", Box::into_raw(Box::new(Aligned(0))) as usize);
}

////////////////////////////////////////////////////////////////
/////////////// below is a more involved example ///////////////
////////////////////////////////////////////////////////////////

#[repr(align(0x100000))]
struct AlignedA(u8);

#[repr(align(0x100000))]
struct AlignedB(u8);

struct FnPtr<T>(fn() -> Box<T>);

struct AG<T>(T);

type AGAlias<T> = AG<T>;

struct AGWithArgs<A, B> {
    not_aligned: FnPtr<A>,
    aligned: AGAlias<B>,
}

static XG_ARGS: AGWithArgs<AlignedA, AlignedB> = AGWithArgs {
    not_aligned: FnPtr(box_aligned),
    aligned: AG(AlignedB(0)),
};

fn box_aligned() -> Box<AlignedA> {
    Box::new(AlignedA(0))
}
