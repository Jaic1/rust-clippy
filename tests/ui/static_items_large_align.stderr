error: this static item (itself or its subfield) has a type alignment,
       which is larger than page size and may not be fulfilled,
       for more information, see <https://github.com/rust-lang/rust/issues/70022>.
  --> $DIR/static_items_large_align.rs:24:1
   |
LL | static X: Aligned = Aligned(0);
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |
   = note: `-D clippy::static-items-large-align` implied by `-D warnings`
note: struct `Aligned` has alignment 0x100000, which is larger than 0x1000,
      if you know what you are doing, config the default page size clippy uses in clippy.toml
  --> $DIR/static_items_large_align.rs:7:1
   |
LL | struct Aligned(u8);
   | ^^^^^^^^^^^^^^

error: this static item (itself or its subfield) has a type alignment,
       which is larger than page size and may not be fulfilled,
       for more information, see <https://github.com/rust-lang/rust/issues/70022>.
  --> $DIR/static_items_large_align.rs:26:1
   |
LL | static X_REF: &Aligned = &Aligned(0);
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |
note: this reference contains an inner type with large alignment
  --> $DIR/static_items_large_align.rs:26:15
   |
LL | static X_REF: &Aligned = &Aligned(0);
   |               ^^^^^^^^
note: struct `Aligned` has alignment 0x100000, which is larger than 0x1000,
      if you know what you are doing, config the default page size clippy uses in clippy.toml
  --> $DIR/static_items_large_align.rs:7:1
   |
LL | struct Aligned(u8);
   | ^^^^^^^^^^^^^^

error: this static item (itself or its subfield) has a type alignment,
       which is larger than page size and may not be fulfilled,
       for more information, see <https://github.com/rust-lang/rust/issues/70022>.
  --> $DIR/static_items_large_align.rs:28:1
   |
LL | static X_PTR: AlignedPtr = AlignedPtr(&Aligned(0) as *const _);
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |
note: struct `AlignedPtr` contains an inner type with large alignment
  --> $DIR/static_items_large_align.rs:18:1
   |
LL | struct AlignedPtr(*const Aligned);
   | ^^^^^^^^^^^^^^^^^
note: struct `Aligned` has alignment 0x100000, which is larger than 0x1000,
      if you know what you are doing, config the default page size clippy uses in clippy.toml
  --> $DIR/static_items_large_align.rs:7:1
   |
LL | struct Aligned(u8);
   | ^^^^^^^^^^^^^^

error: this static item (itself or its subfield) has a type alignment,
       which is larger than page size and may not be fulfilled,
       for more information, see <https://github.com/rust-lang/rust/issues/70022>.
  --> $DIR/static_items_large_align.rs:30:1
   |
LL | static XW: AlignedWrapper = AlignedWrapper { f: 0, g: Aligned(0) };
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |
note: struct `AlignedWrapper` contains an inner type with large alignment
  --> $DIR/static_items_large_align.rs:9:1
   |
LL | struct AlignedWrapper {
   | ^^^^^^^^^^^^^^^^^^^^^
note: struct `Aligned` has alignment 0x100000, which is larger than 0x1000,
      if you know what you are doing, config the default page size clippy uses in clippy.toml
  --> $DIR/static_items_large_align.rs:7:1
   |
LL | struct Aligned(u8);
   | ^^^^^^^^^^^^^^

error: this static item (itself or its subfield) has a type alignment,
       which is larger than page size and may not be fulfilled,
       for more information, see <https://github.com/rust-lang/rust/issues/70022>.
  --> $DIR/static_items_large_align.rs:32:1
   |
LL | static ARRAY: [Aligned; 10] = [Aligned(0); 10];
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |
note: this array contains an inner type with large alignment
  --> $DIR/static_items_large_align.rs:32:15
   |
LL | static ARRAY: [Aligned; 10] = [Aligned(0); 10];
   |               ^^^^^^^^^^^^^
note: struct `Aligned` has alignment 0x100000, which is larger than 0x1000,
      if you know what you are doing, config the default page size clippy uses in clippy.toml
  --> $DIR/static_items_large_align.rs:7:1
   |
LL | struct Aligned(u8);
   | ^^^^^^^^^^^^^^

error: this static item (itself or its subfield) has a type alignment,
       which is larger than page size and may not be fulfilled,
       for more information, see <https://github.com/rust-lang/rust/issues/70022>.
  --> $DIR/static_items_large_align.rs:34:1
   |
LL | static TUPLE: (u8, (Aligned, u8)) = (0, (Aligned(0), 0));
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |
note: this tuple contains an inner type with large alignment
  --> $DIR/static_items_large_align.rs:34:15
   |
LL | static TUPLE: (u8, (Aligned, u8)) = (0, (Aligned(0), 0));
   |               ^^^^^^^^^^^^^^^^^^^
note: struct `Aligned` has alignment 0x100000, which is larger than 0x1000,
      if you know what you are doing, config the default page size clippy uses in clippy.toml
  --> $DIR/static_items_large_align.rs:7:1
   |
LL | struct Aligned(u8);
   | ^^^^^^^^^^^^^^

error: this static item (itself or its subfield) has a type alignment,
       which is larger than page size and may not be fulfilled,
       for more information, see <https://github.com/rust-lang/rust/issues/70022>.
  --> $DIR/static_items_large_align.rs:36:1
   |
LL | static XE: AlignedEnum = AlignedEnum::A(Aligned(0));
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |
note: enum `AlignedEnum` contains an inner type with large alignment
  --> $DIR/static_items_large_align.rs:14:1
   |
LL | enum AlignedEnum {
   | ^^^^^^^^^^^^^^^^
note: struct `Aligned` has alignment 0x100000, which is larger than 0x1000,
      if you know what you are doing, config the default page size clippy uses in clippy.toml
  --> $DIR/static_items_large_align.rs:7:1
   |
LL | struct Aligned(u8);
   | ^^^^^^^^^^^^^^

error: this static item (itself or its subfield) has a type alignment,
       which is larger than page size and may not be fulfilled,
       for more information, see <https://github.com/rust-lang/rust/issues/70022>.
  --> $DIR/static_items_large_align.rs:38:1
   |
LL | static XG: AlignedGeneric<Aligned> = AlignedGeneric(Aligned(0));
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |
note: struct `AlignedGeneric` with substitutions [
          Aligned,
      ],
      contains an inner type with large alignment
  --> $DIR/static_items_large_align.rs:22:1
   |
LL | struct AlignedGeneric<T>(T);
   | ^^^^^^^^^^^^^^^^^^^^^^^^
note: struct `Aligned` has alignment 0x100000, which is larger than 0x1000,
      if you know what you are doing, config the default page size clippy uses in clippy.toml
  --> $DIR/static_items_large_align.rs:7:1
   |
LL | struct Aligned(u8);
   | ^^^^^^^^^^^^^^

error: this static item (itself or its subfield) has a type alignment,
       which is larger than page size and may not be fulfilled,
       for more information, see <https://github.com/rust-lang/rust/issues/70022>.
  --> $DIR/static_items_large_align.rs:40:1
   |
LL | static XGW: AlignedGeneric<AlignedWrapper> = AlignedGeneric(AlignedWrapper { f: 0, g: Aligned(0) });
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |
note: struct `AlignedGeneric` with substitutions [
          AlignedWrapper,
      ],
      contains an inner type with large alignment
  --> $DIR/static_items_large_align.rs:22:1
   |
LL | struct AlignedGeneric<T>(T);
   | ^^^^^^^^^^^^^^^^^^^^^^^^
note: struct `AlignedWrapper` contains an inner type with large alignment
  --> $DIR/static_items_large_align.rs:9:1
   |
LL | struct AlignedWrapper {
   | ^^^^^^^^^^^^^^^^^^^^^
note: struct `Aligned` has alignment 0x100000, which is larger than 0x1000,
      if you know what you are doing, config the default page size clippy uses in clippy.toml
  --> $DIR/static_items_large_align.rs:7:1
   |
LL | struct Aligned(u8);
   | ^^^^^^^^^^^^^^

error: this static item (itself or its subfield) has a type alignment,
       which is larger than page size and may not be fulfilled,
       for more information, see <https://github.com/rust-lang/rust/issues/70022>.
  --> $DIR/static_items_large_align.rs:68:1
   |
LL | / static XG_ARGS: AGWithArgs<AlignedA, AlignedB> = AGWithArgs {
LL | |     not_aligned: FnPtr(box_aligned),
LL | |     aligned: AG(AlignedB(0)),
LL | | };
   | |__^
   |
note: struct `AGWithArgs` with substitutions [
          AlignedA,
          AlignedB,
      ],
      contains an inner type with large alignment
  --> $DIR/static_items_large_align.rs:63:1
   |
LL | struct AGWithArgs<A, B> {
   | ^^^^^^^^^^^^^^^^^^^^^^^
note: struct `AG` with substitutions [
          AlignedB,
      ],
      contains an inner type with large alignment
  --> $DIR/static_items_large_align.rs:59:1
   |
LL | struct AG<T>(T);
   | ^^^^^^^^^^^^
note: struct `AlignedB` has alignment 0x100000, which is larger than 0x1000,
      if you know what you are doing, config the default page size clippy uses in clippy.toml
  --> $DIR/static_items_large_align.rs:55:1
   |
LL | struct AlignedB(u8);
   | ^^^^^^^^^^^^^^^

error: aborting due to 10 previous errors

